src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,545 @@
+    public int[] getRequiredTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        variables = new ArrayDeque<>();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        switch (ast.getType()) {
+            case TokenTypes.DOT:
+                final DetailAST firstDotOrMethodCallAst = getFirstDotOrMethodCallToken(ast);
+                if (firstDotOrMethodCallAst == null
+                        && ast.findFirstToken(TokenTypes.LITERAL_NEW) == null
+                        && ast.findFirstToken(TokenTypes.LITERAL_SUPER) == null
+                        && ast.findFirstToken(TokenTypes.LITERAL_CLASS) == null
+                        && ast.findFirstToken(TokenTypes.LITERAL_THIS) == null) {
+                    checkIdentifierAst(ast.findFirstToken(TokenTypes.IDENT));
+                }
+                break;
+            case TokenTypes.VARIABLE_DEF:
+                if (ast.getParent().getType() != TokenTypes.OBJBLOCK) {
+                    final DetailAST ident = ast.findFirstToken(TokenTypes.IDENT);
+                    final VariableDesc desc = new VariableDesc(ident.getText(), ast);
+                    desc.setScopeOfVariable(ast.getParent());
+                    variables.push(desc);
+                }
+                break;
+            case TokenTypes.IDENT:
+                if (!TokenUtil.isOfType(ast.getParent().getType(),
+                        TokenTypes.VARIABLE_DEF, TokenTypes.DOT,
+                        TokenTypes.LITERAL_NEW, TokenTypes.PATTERN_VARIABLE_DEF)) {
+                    checkIdentifierAst(ast);
+                }
+                break;
+            case TokenTypes.SLIST:
+            case TokenTypes.EXPR:
+            case TokenTypes.ELIST:
+                break;
+            case TokenTypes.LITERAL_INSTANCEOF:
+                final DetailAST patternVariableDef = ast
+                        .findFirstToken(TokenTypes.PATTERN_VARIABLE_DEF);
+                if (patternVariableDef != null) {
+                    final DetailAST identToken = patternVariableDef
+                            .findFirstToken(TokenTypes.IDENT);
+                    final VariableDesc variableDesc =
+                            new VariableDesc(identToken.getText(), patternVariableDef);
+                    variableDesc.registerAsPatternVariable();
+                    setScopeOfPatternVariable(ast, patternVariableDef, variableDesc);
+                    variables.push(variableDesc);
+                }
+                break;
+            default:
+                throw new IllegalStateException(ast.toString());
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        if (TokenUtil.isOfType(ast.getType(), TokenTypes.SLIST, TokenTypes.EXPR,
+                TokenTypes.ELIST)) {
+            int size = variables.size();
+            while (size > 0 && variables.peek().getScopeOfVariable() == ast) {
```

Comment:
`size > 0` better to use `!variables.isEmpty()`

Reply:
done